crates/common/src/identity.rs

```diff
@@ -1,29 +1,116 @@
+use std::fmt;
+
 use crate::{
     prov::AgentId,
     signing::{DirectoryStoredKeys, SignerError},
 };
 
 use k256::ecdsa::{signature::Signer, Signature, SigningKey, VerifyingKey};
 use thiserror::Error;
+use tracing::warn;
 
 #[derive(Error, Debug)]
 pub enum IdentityError {
-    #[error("Invalid key store directory {0}")]
+    #[error("Failed to update context")]
+    OpaContext,
+
+    #[error("Failed to get agent id from JWT claims")]
+    JwtClaims,
+
+    #[error("Invalid key store directory: {0}")]
     KeyStore(#[from] SignerError),
-    #[error("Malformed JSON {0}")]
+
+    #[error("Malformed JSON: {0}")]
     SerdeJson(#[from] serde_json::Error),
-    #[error("Serialization error {0}")]
+
+    #[error("Serialization error: {0}")]
     SerdeJsonSerialize(String),
-    #[error("Signing error {0}")]
+
+    #[error("Signing error: {0}")]
     Signing(#[from] k256::ecdsa::Error),
 }
 
+/// Claims from a JWT, referenced in creating an AgentId for a Chronicle user
+#[derive(Clone, Debug, Deserialize, Serialize)]
+pub struct JwtClaims(pub serde_json::Map<String, serde_json::Value>);
+
+/// Contextual data for OPA from a request via GraphQL
+pub struct GraphQlOpaContext {
+    parent_type: serde_json::Value,
+    resolve_path: serde_json::Value,
+    identity_claims: serde_json::Value,
+}
+
+impl GraphQlOpaContext {
+    pub fn new(
+        parent_type: serde_json::Value,
+        resolve_path: serde_json::Value,
+        identity_claims: serde_json::Value,
+    ) -> Self {
+        Self {
+            parent_type,
+            resolve_path,
+            identity_claims,
+        }
+    }
+}
+
+/// Contextual data for OPA in the Transaction Processor
+#[derive(Debug)]
+pub struct TpOpaContext {
+    identity: serde_json::Value,
+    operation: serde_json::Value,
+    state: Vec<serde_json::Value>,
+}
+
+impl TpOpaContext {
+    pub fn new(
+        identity: serde_json::Value,
+        operation: serde_json::Value,
+        state: Vec<serde_json::Value>,
+    ) -> Self {
+        Self {
+            identity,
+            operation,
+            state,
+        }
+    }
+}
+
+/// Contains the scalar ID and OPA contextual data for a user established via JWT
+#[derive(Serialize, Deserialize, PartialEq, Eq, Debug, Clone)]
+pub struct JwtId {
+    pub id: AgentId,
+    #[serde(flatten)]
+    pub context: serde_json::Value,
+}
+
+impl JwtId {
+    fn new(external_id: &str, context: serde_json::Value) -> Self {
+        Self {
+            id: AgentId::from_external_id(external_id),
+            context,
+        }
+    }
+}
+
+/// Chronicle identity object for authorization
 #[derive(Serialize, Deserialize, PartialEq, Eq, Debug, Clone)]
-#[serde(tag = "type", content = "id")]
+#[serde(tag = "type")]
 pub enum AuthId {
-    Anonymous,
-    Chronicle,
-    JWT(AgentId),
+    Anonymous(serde_json::Value),
```

Comment:
Curious what these Values are for. Do we ever use them?

Reply:
Is your sense that context data for OPA checks is irrelevant in all cases to Chronicle and Anonymous users?

Reply:
I just couldn't find where we put any actual context data in for those users but I may have just missed the code.

Reply:
Context is the WHAT of an operation, Identity is the WHO. 

The WHO will be the same in both graphql and the transaction processor but the WHAT is different - in graphql it's whatever async_graphql thinks we are trying to do via context, in the transaction processor it is the chronicle operation (which  has already passed and doesn't need to be part of the WHO).

Reply:
So here anonymous is just anonymous, the context is separate as it is dependent on where it is being checked.